### PR TITLE
Grids GUI test

### PR DESCRIPTION
**Pull request by scottyak**
This is just a skeleton to demonstrate how the UI test would work. More cleanups to follow if folks agree on the general approach.

### CODE REVIEW

**Comment by youngmit** [comment]
Looking good so far. Thanks for turning the crank on Travis; truly a thankless job :(

**Comment by scottyak** [comment]
Thanks!

Here's the progress I've made so far:
- Found a way to include wxpython as a dependency in one of the Travis environments. Building the wheel from source would have taken ~30mins, but thankfully there is a prebuilt wheel for xenial which we can just download, so wxpython only takes ~1 min to install.
- Headless UI test works in my local Ubuntu environment -- runs in seconds, and the test passes.
- Headless UI test runs in the Travis environment without crashing, but the test fails (even though the code is the same as the one which passed in my local environment.) "Works on my machine", indeed.

So my current problem is getting the test to pass in the Travis environment. This is not easy to do, because there isn't a display that allows me to see what the test is doing to the UI.

I have two approaches in mind, both of which I cannot do by myself:

1. Video capture the virtual display when the test is running, and store the video in some place I can access.
    * How to record the screen:
        * Use PIL, opencv, and maybe FFMPEG to implement screen recording functionality in the GridGuiTestCase class, or
        * Install [byzanz](https://linux.die.net/man/1/byzanz-record), and use before_script and after_script to start and stop the screen recording.
    * How to upload the video:
        * There is some documentation [here](https://docs.travis-ci.com/user/uploading-artifacts/) for how to upload artifacts from Travis to S3, but it would require someone from Terrapower to configure S3 access keys, and grant me access to some of the artifacts.
1. Requesting for remote SSH debugging access, and set up X11 forwarding to see what's going on. Unfortunately I am unable to find much information about how exactly to do this.

I am currently experimenting with the screen recording approach, and testing locally. Will let you know how this goes!

**Comment by youngmit** [comment]
Sounds like you've made lots of progress. I think having tests that can run in _any_ environment is enough of an improvement that it'd a shame to hold this up getting the tests to run under Travis. Maybe for now we can just disable the test under Travis, merge this PR and go from there?

Looks like a good way to disable tests under travis is to set an environment variable in the travis config and look for it in a `skipif` clause in the test (https://stackoverflow.com/questions/57333637/how-i-can-disable-some-tests-in-travis-ci-for-python)

**Comment by youngmit** [comment]
So I think path forward would be to revert most of the Travis changes, but probably retain them on another branch somewhere; I think we will want to circle back to this, and it looks like you learned a lot. Let's add an env var to disable problematic tests. We can close #133 and open a new issue for getting the tests to work under Travis.

**Comment by scottyak** [comment]
Sounds good to me!

**Review by youngmit — COMMENTED** [review]
It looks like there is still stuff left over from trying to figure out what Travis is up to (sorry, i know its _really_ frustrating to try to figure out what CI systems are up to, especially when `pytest` is so eager to hide all output from you 😡). So let me know when you think this is good to go and I'll do a full review.

For now, I want to get ahead of some content things earlier rather than later. Mainly, I've got to look into whether we are okay with the little LGPL snippet. I think for that to be there, we would need to have a license notice displayed somewhat prominently that there is LGPL code in ARMI. Not sure if we want to go that route (at least not to accommodate such a little snippet). Also, _technically_ according to the CLA, anything that isn't original work needs to be submitted separately (i think so that we can consider it's copyright holder and license terms separately and make sure we incorporate everything properly). So! if at all possible, it would be nice if the `_cleanup()` function could be adjusted/adapted enough that we can make a fair-use argument that it is original work. Think that's possible?

**Comment by scottyak** [comment]
Yep, happy to adjust the _cleanup() function. I agree that it's best to avoid having LGPL code in the codebase.

Will update once the PR is in a more ready state for a proper review!

File: armi/utils/gridEditor.py
Comment (youngmit) [inline armi/utils/gridEditor.py:1062]: Bumped into this in the course of my work today, fixing in #236; apologies for the conflict to come

**Comment by scottyak** [comment]
Ready for review, PTAL! Thanks!

File: armi/utils/gridEditor.py
Comment (scottyak) [reply armi/utils/gridEditor.py]: No prob!

**Comment by scottyak** [comment]
Oops, not quite ready yet! Will update again when ready.

**Comment by scottyak** [comment]
This is ready now!

**Comment by scottyak** [comment]
(Not sure why test coverage dropped.)

**Comment by ntouran** [comment]
Thanks for all this work. I'm not sure why Travis test results are so detached from this PR in general but checking into it, it looks like the unguarded `import wx` in `test_gridGui.py` is [causing an import error in the normal unit test suite](https://travis-ci.org/github/terrapower/armi/jobs/754815044). Then most of the other tests didn't run so coverage is artificially low. 

Maybe shove the `import wx` into the environment conditional as well?

**Comment by scottyak** [comment]
Now there appears to be an issue with "pip install black".

**Comment by scottyak** [comment]
It's ready now!

**Comment by ntouran** [comment]
Whew, nice. 

Though our test suite only runs up to Python 3.7, I ran this on my Python 3.8 Linux box and got a failure. Have you seen anything like this? Could indicate that there's some important change in an underlying lib that we don't have expressed in requirements.txt, etc.?

```
$ pytest --no-xvfb --capture=tee-sys test_gridGui.py
Initializing generic ARMI Framework application

                       ---------------------------------------------------
                      |             _      ____     __  __    ___         |
                      |            / \    |  _ \   |  \/  |  |_ _|        |
                      |           / _ \   | |_) |  | |\/| |   | |         |
                      |          / ___ \  |  _ <   | |  | |   | |         |
                      |         /_/   \_\ |_| \_\  |_|  |_|  |___|        |
                      |         Advanced  Reactor  Modeling Interface     |
                       ---------------------------------------------------
                                      Version 0.1.6     

=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.8.6, pytest-5.4.3, py-1.8.1, pluggy-0.13.1
rootdir: /home/nick/code/armi, inifile: pytest.ini
plugins: xvfb-2.0.0, cov-2.9.0, html-2.1.1, xdist-1.32.0, metadata-1.9.0, forked-1.1.3
collected 1 item                                                                                                                                                                                                  

test_gridGui.py 10:44:09 AM: Debug: ClientToScreen cannot work when toplevel window is not shown
110:44:09 AM: Debug: ClientToScreen cannot work when toplevel window is not shown
10:44:09 AM: Debug: ClientToScreen cannot work when toplevel window is not shown
10:44:09 AM: Debug: ClientToScreen cannot work when toplevel window is not shown
F                                                                                                                                                                                           [100%]

==================================================================================================== FAILURES =====================================================================================================
______________________________________________________________________________________________ Test.test_setNumRings ______________________________________________________________________________________________
Traceback (most recent call last):
  File "/home/nick/code/armi/armi/utils/tests/test_gridGui.py", line 159, in test_setNumRings
    self.assertEqual(1, len(gridCellIndices))
  File "/usr/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 1 != 196
================================================================================================ warnings summary =================================================================================================
armi/utils/tests/test_gridGui.py::Test::test_setNumRings
armi/utils/tests/test_gridGui.py::Test::test_setNumRings
armi/utils/tests/test_gridGui.py::Test::test_setNumRings
armi/utils/tests/test_gridGui.py::Test::test_setNumRings
  /home/nick/code/armi/armi/utils/gridEditor.py:1245: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    return wx.Rect(wx.Point(*topLeft[:2]), wx.Point(*bottomRight[:2]))

armi/utils/tests/test_gridGui.py: 784 tests with warnings
  /home/nick/code/armi/armi/utils/gridEditor.py:163: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    return wx.Rect(wx.Point(xmin, ymin), wx.Point(xmax, ymax))

armi/utils/tests/test_gridGui.py: 1568 tests with warnings
  /home/nick/code/armi/armi/utils/gridEditor.py:261: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    dc.DrawPolygon(poly)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
============================================================================================ slowest 30 test durations ============================================================================================
1.43s call     armi/utils/tests/test_gridGui.py::Test::test_setNumRings

(0.00 durations hidden.  Use -vv to show these durations.)
============================================================================================= short test summary info =============================================================================================
FAILED test_gridGui.py::Test::test_setNumRings - AssertionError: 1 != 196
```

**Comment by scottyak** [comment]
That's unfortunate -- looks like it only works on my machine, eh?

This is what I see:
```
$ pytest -vv --no-xvfb --capture=tee-sys armi/utils/tests/test_gridGui.py 
Initializing generic ARMI Framework application

                       ---------------------------------------------------
                      |             _      ____     __  __    ___         |
                      |            / \    |  _ \   |  \/  |  |_ _|        |
                      |           / _ \   | |_) |  | |\/| |   | |         |
                      |          / ___ \  |  _ <   | |  | |   | |         |
                      |         /_/   \_\ |_| \_\  |_|  |_|  |___|        |
                      |         Advanced  Reactor  Modeling Interface     |
                       ---------------------------------------------------
                                      Version 0.1.6     

============================================================================================= test session starts =============================================================================================
platform linux -- Python 3.8.5, pytest-6.2.1, py-1.10.0, pluggy-0.13.1 -- /home/scott/armi-venv/bin/python3.8
cachedir: .pytest_cache
metadata: {'Python': '3.8.5', 'Platform': 'Linux-5.8.0-38-generic-x86_64-with-glibc2.29', 'Packages': {'pytest': '6.2.1', 'py': '1.10.0', 'pluggy': '0.13.1'}, 'Plugins': {'xvfb': '2.0.0', 'metadata': '1.11.0'}}
rootdir: /home/scott/armi-venv/armi, configfile: pytest.ini
plugins: xvfb-2.0.0, metadata-1.11.0
collected 1 item                                                                                                                                                                                              

armi/utils/tests/test_gridGui.py::Test::test_setNumRings PASSED                                                                                                                                         [100%]

============================================================================================== warnings summary ===============================================================================================
armi/utils/tests/test_gridGui.py: 12 warnings
  /home/scott/armi-venv/armi/armi/utils/gridEditor.py:1245: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    return wx.Rect(wx.Point(*topLeft[:2]), wx.Point(*bottomRight[:2]))

armi/utils/tests/test_gridGui.py: 1572 warnings
  /home/scott/armi-venv/armi/armi/utils/gridEditor.py:163: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    return wx.Rect(wx.Point(xmin, ymin), wx.Point(xmax, ymax))

armi/utils/tests/test_gridGui.py: 3144 warnings
  /home/scott/armi-venv/armi/armi/utils/gridEditor.py:261: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    dc.DrawPolygon(poly)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================ slowest 30 durations =============================================================================================
1.21s call     armi/utils/tests/test_gridGui.py::Test::test_setNumRings
0.00s setup    armi/utils/tests/test_gridGui.py::Test::test_setNumRings
0.00s teardown armi/utils/tests/test_gridGui.py::Test::test_setNumRings
====================================================================================== 1 passed, 4728 warnings in 1.32s =======================================================================================
```

I'm wondering if it could be some difference between our display settings. Would you mind running the test again without disabling xvfb and pasting the output here? Like this:
```
$ pytest -vv --capture=tee-sys armi/utils/tests/test_gridGui.py
```

**Comment by ntouran** [comment]
The previous time I was running through a ssh with X forwarding so I thought that might be an issue and ran again locally with this command. Similarly, I get:

```
(py38) nick@lemur:~/code/armi$ pytest -vv --capture=tee-sys armi/utils/tests/test_gridGui.py
Initializing generic ARMI Framework application

                       ---------------------------------------------------
                      |             _      ____     __  __    ___         |
                      |            / \    |  _ \   |  \/  |  |_ _|        |
                      |           / _ \   | |_) |  | |\/| |   | |         |
                      |          / ___ \  |  _ <   | |  | |   | |         |
                      |         /_/   \_\ |_| \_\  |_|  |_|  |___|        |
                      |         Advanced  Reactor  Modeling Interface     |
                       ---------------------------------------------------
                                      Version 0.1.6     

=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.8.6, pytest-6.2.1, py-1.10.0, pluggy-0.13.1 -- /opt/venvs/py38/bin/python3
cachedir: .pytest_cache
metadata: {'Python': '3.8.6', 'Platform': 'Linux-5.8.0-7630-generic-x86_64-with-glibc2.32', 'Packages': {'pytest': '6.2.1', 'py': '1.10.0', 'pluggy': '0.13.1'}, 'Plugins': {'html': '3.1.1', 'xvfb': '2.0.0', 'xdist': '2.2.0', 'metadata': '1.11.0', 'cov': '2.10.1', 'forked': '1.3.0'}}
rootdir: /home/nick/code/armi, configfile: pytest.ini
plugins: html-3.1.1, xvfb-2.0.0, xdist-2.2.0, metadata-1.11.0, cov-2.10.1, forked-1.3.0
collected 1 item                                                                                                                                                                                                  

armi/utils/tests/test_gridGui.py::Test::test_setNumRings FAILED                                                                                                                                             [100%]

==================================================================================================== FAILURES =====================================================================================================
______________________________________________________________________________________________ Test.test_setNumRings ______________________________________________________________________________________________
Traceback (most recent call last):
  File "/home/nick/code/armi/armi/utils/tests/test_gridGui.py", line 159, in test_setNumRings
    self.assertEqual(1, len(gridCellIndices))
  File "/usr/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python3.8/unittest/case.py", line 905, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 1 != 196
================================================================================================ warnings summary =================================================================================================
armi/utils/tests/test_gridGui.py: 12 warnings
  /home/nick/code/armi/armi/utils/gridEditor.py:1245: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    return wx.Rect(wx.Point(*topLeft[:2]), wx.Point(*bottomRight[:2]))

armi/utils/tests/test_gridGui.py: 1572 warnings
  /home/nick/code/armi/armi/utils/gridEditor.py:163: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    return wx.Rect(wx.Point(xmin, ymin), wx.Point(xmax, ymax))

armi/utils/tests/test_gridGui.py: 3144 warnings
  /home/nick/code/armi/armi/utils/gridEditor.py:261: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
    dc.DrawPolygon(poly)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================== slowest 30 durations ===============================================================================================
1.07s call     armi/utils/tests/test_gridGui.py::Test::test_setNumRings
0.00s setup    armi/utils/tests/test_gridGui.py::Test::test_setNumRings
0.00s teardown armi/utils/tests/test_gridGui.py::Test::test_setNumRings
============================================================================================= short test summary info =============================================================================================
FAILED armi/utils/tests/test_gridGui.py::Test::test_setNumRings - AssertionError: 1 != 196
```

**Comment by youngmit** [comment]
@scottyak I am going to try to get these tests working locally myself, and once successful, I think we should merge this. You've done a lot of great work on this, and we can always circle back to this long-tail stuff later. Having any tests at all is going to be a huge get, even if it's not running under CI.

**Comment by youngmit** [comment]
If you don't mind me asking, what sort of environment are you running under locally? Distro, python version, etc? If i get stuck, a `pip freeze` output might be useful as well.

File: armi/utils/tests/test_gridGui.py
Comment (youngmit) [reply armi/utils/tests/test_gridGui.py:29]: In order to get the tests to run, I had to be in a pure X11 environment; default Gnome on my system runs on Wayland with xwayland to support X clients, which didn't work. Now that I think about it, the more security-minded nature of Wayland may even make this sort of GUI testing impossible? I'm not sure... anyways, it might be worth noting known requirements on the test environment.

File: armi/utils/tests/test_gridGui.py
Comment (youngmit) [inline armi/utils/tests/test_gridGui.py:118]: Since these tests are a bit touchy, we may just want to make them opt-in?
Something like this seems to work well:
```
@pytest.mark.skipif(not bool(os.environ.get("ARMI_GUI_TESTS", False)), reason="GUI tests require a rather specific environment (see above), so these tests are opt-in")
```

File: tox.ini
Comment (youngmit) [inline tox.ini:12]: If we use the environment variable-based `skipif` thingy, we can disable the GUI tests by deglecting to set the `ARMI_GUI_TESTS` variable

File: .coveragerc
Comment (youngmit) [inline .coveragerc:12]: Even with the `ARMI_GUI_TESTS` variable thing, we should leave these so the coverage report doesnt make us too sad

File: armi/utils/tests/test_gridGui.py
Comment (youngmit) [inline armi/utils/tests/test_gridGui.py:43]: This ended up being a bit short for my system; raising it to 0.05 helped for me.

**Review by youngmit — CHANGES_REQUESTED** [review]
I was able to make this work on my end! I had to make sure that i was running under a pure X11 environment. On my system, the default Gnome DE runs under Wayland. If you think you have this running under Wayland, let me know! Since this might go over some folks' heads, I'm thinking that we may want to make the tests opt-in. I played around with a `@pytest.mark.skipif` decorator, which seemed to do the trick. I also had to make the tick time a little longer for the assertion to reliably pass; i think it was asserting before the app had a chance to fully respond to the event.

So i think that `skipif` and a longer tick, and this is good to merge. Thanks so much! I know fighting with Travis is a bit of a slog :(

File: armi/utils/tests/test_gridGui.py
Comment (scottyak) [reply armi/utils/tests/test_gridGui.py:43]: Thanks, raised to 0.05.

**Comment by scottyak** [comment]
> If you don't mind me asking, what sort of environment are you running under locally? Distro, python version, etc? If i get stuck, a `pip freeze` output might be useful as well.

~/armi-venv/armi$ pip freeze
apipkg==1.5
appdirs==1.4.4
-e git+git@github.com:scottyak/armi.git@17664097b69774d5adefee0259d8500b1cce97e5#egg=armi
attrs==20.3.0
black==20.8b1
click==7.1.2
configparser==5.0.1
coverage==5.3.1
cycler==0.10.0
data-science-types==0.2.22
EasyProcess==0.3
execnet==1.7.1
future==0.18.2
h5py==2.10.0
iniconfig==1.1.1
kiwisolver==1.3.1
matplotlib==3.3.3
mypy==0.790
mypy-extensions==0.4.3
numpy==1.19.3
opencv-python==4.5.1.48
ordered-set==4.0.2
packaging==20.8
pathspec==0.8.1
Pillow==8.0.1
pluggy==0.13.1
psutil==5.8.0
py==1.10.0
pyevtk==1.1.1
Pympler==0.9
pyparsing==2.4.7
pytest==6.2.1
pytest-metadata==1.11.0
pytest-xvfb==2.0.0
python-config==0.1.2
python-dateutil==2.8.1
PyVirtualDisplay==2.0
PyYAML==5.3.1
regex==2020.11.13
ruamel.yaml==0.16.12
ruamel.yaml.clib==0.2.2
scipy==1.6.0
screeninfo==0.6.7
six==1.15.0
tabulate==0.8.7
toml==0.10.2
typed-ast==1.4.2
typing-extensions==3.7.4.3
voluptuous==0.12.1
wxPython==4.1.1
xlrd==2.0.1
yamlize==0.7.0

File: .coveragerc
Comment (scottyak) [reply .coveragerc:12]: 👍

File: armi/utils/tests/test_gridGui.py
Comment (scottyak) [reply armi/utils/tests/test_gridGui.py:118]: Done.

File: armi/utils/tests/test_gridGui.py
Comment (scottyak) [reply armi/utils/tests/test_gridGui.py:29]: Thanks for figuring this out! I've noted this in the comment.

File: tox.ini
Comment (scottyak) [reply tox.ini:12]: A little worried whether the `import wx` statement would crash the test in Travis, I'll try it with and without the ignore statements, and see if we can do without them.

Comment (scottyak) [reply tox.ini:12]: It appears that we can't do without the `--ignore` flags:
- With `--ignore`: https://coveralls.io/builds/36415490
- Without `--ignore`: https://coveralls.io/builds/36415610

Comment (youngmit) [reply tox.ini:12]: yeah, no kidding. okay, we can stomp on this later

**Review by youngmit — APPROVED** [review]
Alrighty, this looks good to go! Thanks for doing this :) Really good to have some coverage on this GUI stuff.

**Comment by scottyak** [comment]
And thank you all for being so responsive to comments!

**Pull request merged**
